tokio/src/net/tcp/stream.rs

```diff
@@ -915,6 +915,138 @@ impl TcpStream {
+    ///
+    /// If the socket is ready for writing, the provided closure is called. The
+    /// closure should attempt to write from the socket by manually calling the
+    /// appropriate syscall. If the operation fails because the socket is not
+    /// actually ready, then the closure should return a `WouldBlock` error and
+    /// the write readiness flag is cleared. The return value of the closure is
+    /// then returned by `try_write_io`.
+    ///
+    /// If the socket is not ready for writing, then the closure is not called
+    /// and a `WouldBlock` error is returned.
+    ///
+    /// The closure should only return a `WouldBlock` error if it has performed
+    /// an IO operation on the socket that failed due to the socket not being
+    /// ready. Returning a `WouldBlock` error in any other situation will
+    /// incorrectly clear the readiness flag, which can cause the socket to
+    /// behave incorrectly.
+    ///
+    /// The closure should not perform the write operation using any of the
+    /// methods defined on the Tokio `TcpStream` type, as this will mess with
+    /// the readiness flag and can cause the socket to behave incorrectly.
+    ///
+    /// Usually, [`writable()`] or [`ready()`] is used with this function.
+    ///
+    /// [`writable()`]: TcpStream::writable()
+    /// [`ready()`]: TcpStream::ready()
+    pub fn try_write_io<R>(&self, f: impl FnOnce() -> io::Result<R>) -> io::Result<R> {
+        self.io.registration().try_io(Interest::WRITABLE, f)
+    }
+
+    /// Polls for write readiness and then calls the `f` for writing operation.
+    ///
+    /// If the tcp stream is not currently ready for writing, this method will
+    /// store a clone of the `Waker` from the provided `Context`. When the tcp
+    /// stream becomes ready for writing, `Waker::wake` will be called on the
+    /// waker.
+    ///
+    /// Note that on multiple calls to `poll_write_io` only
+    /// the `Waker` from the `Context` passed to the most recent call is
+    /// scheduled to receive a wakeup. (However, `poll_read_io` retains a
+    /// second, independent waker.)
+    ///
+    /// This function is intended for cases where customized I/O operations
+    /// may change the readiness of the underlying socket.
+    /// If the `f` function returns an error `WouldBlock`, then the write
+    /// readiness will be cleared and returns `Poll::Pending`.
+    ///
+    /// # Return value
+    ///
+    /// The function returns:
+    ///
+    /// * `Poll::Pending` if the tcp stream is not ready for writing.
+    /// * `Poll::Ready(Ok(R))` if the `f` returns `Ok(R)`.
+    /// * `Poll::Ready(Err(e))` if an error is encountered from `f` except `WouldBlock`.
+    ///
+    /// # Errors
+    ///
+    /// This function may encounter any standard I/O error except `WouldBlock`.
+    pub fn poll_write_io<R>(
+        &self,
+        cx: &mut Context<'_>,
+        f: impl FnMut() -> io::Result<R>,
```

Comment:
All of the changes I suggested for `poll_read_io` shouls also be applied to `poll_write_io`.